File: 90.子集-ii.cpp
/*
 * @lc app=leetcode.cn id=90 lang=cpp
 *
 * [90] 子集 II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        set<vector<int>> tmp;
        vector<int> current;
        foo(nums, 0, current, tmp);
        vector<vector<int>> res;
        for (const auto& vec : tmp) {
            res.push_back(vec);
        }
        return res;
    } 
    void foo(const vector<int>& nums, int idx, 
                vector<int>& current, set<vector<int>>& res) {
        std::sort(current.begin(), current.end());
        res.insert(current);
        for (int i = idx; i < nums.size(); ++i) {
            vector<int> next = current;
            next.push_back(nums[i]);
            foo(nums, i + 1, next, res);
        }
    }
};
// @lc code=end

File: 90.子集-ii.py
#
# @lc app=leetcode.cn id=90 lang=python
#
# [90] 子集 II
#

# @lc code=start
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        sorted_res = []
        def count(start, num):  # 深度优先搜索
            if sorted(num[:]) not in sorted_res:
                res.append(num)
                sorted_res.append(sorted(num[:]))
            for i in range(start, len(nums)):
                count(i + 1, num + [nums[i]])
        count(0, [])
        return res
# @lc code=end

